File: easysocketPy/socket_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import json
import socket
import logging
import traceback

logging.basicConfig(format='%(asctime)s %(message)s', level=logging.DEBUG, filename='log.log')


class SocketClient(object):

    def __init__(self, name, host, port, callback=None):
        if callback is None:
            callback = {}
        self.m_client_socket = None
        self.m_name = name
        self.m_callback = callback
        try:
            self.m_client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.m_client_socket.connect((host, port))
            self.normal_log("连接服务器成功: " + self.receive_data())
        except Exception as err:
            self.error_log("连接服务器失败")
            self.m_client_socket = None

    def error_log(self, title):
        logging.error("[" + self.m_name + "]> " + title + ": " + traceback.format_exc())

    def normal_log(self, text):
        logging.info("[" + self.m_name + "]> " + text)

    def login(self):
        if self.m_client_socket is not None:
            try:
                data = {"name": self.m_name, "action": "login"}
                send_str = json.dumps(data)
                self.m_client_socket.send(bytes(send_str, encoding='utf8'))
                self.normal_log("登录")
                receive = self.receive_data()
                """对应请求的回调函数，如果有就执行一下"""
                if data["action"] in self.m_callback:
                    resultData = json.loads(str(receive, encoding="utf-8"))
                    self.m_callback[data["action"]](resultData)
                return True
            except:
                self.error_log("登录错误")
                self.m_client_socket.shutdown(socket.SHUT_RDWR)
                self.m_client_socket.close()
        return False

    def send_message(self, to_client, message_json):
        result = ""
        if self.m_client_socket is not None:
            try:
                data = {"name": self.m_name, "action": "send", "to": to_client, "data": message_json}
                send_str = json.dumps(data)
                self.m_client_socket.send(bytes(send_str, encoding='utf8'))
                self.normal_log("发送消息")
                result = self.receive_data()
            except:
                self.error_log("发送消息错误")
                self.m_client_socket.shutdown(socket.SHUT_RDWR)
                self.m_client_socket.close()
        return result

    def listen(self):
        if self.m_client_socket is not None:
            try:
                while True:
                    data = {"name": self.m_name, "action": "listen"}
                    send_str = json.dumps(data)
                    self.m_client_socket.send(bytes(send_str, encoding='utf8'))
                    self.normal_log("发出监听请求: " + send_str)
                    receive = self.receive_data()
                    """对应请求的回调函数，如果有就执行一下"""
                    if data["action"] in self.m_callback:
                        resultData = json.loads(str(receive, encoding="utf-8"))
                        self.normal_log("监听消息接收: " + receive)
                        self.m_callback[data["action"]](resultData)
            except:
                self.error_log("监听消息错误")
                self.m_client_socket.shutdown(socket.SHUT_RDWR)
                self.m_client_socket.close()

    def close_connection(self):
        if self.m_client_socket is not None:
            self.m_client_socket.close()
            self.m_client_socket = None
            self.normal_log("客户端主动断开连接")

    def receive_data(self):
        result = ""
        buffer = []
        try:
            buffer = self.m_client_socket.recv(2048)
            result = str(buffer, encoding='utf-8')
            self.normal_log("接收服务器消息: " + result)
        except:
            self.error_log("接收消息错误")
        return result

File: easysocket/SocketClient.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Text;
using Newtonsoft.Json.Linq;

namespace easysocket
{
    public class SocketClient
    {
        public SocketClient(string name, string serverIPAddress, int port)
        {
            Name = name;
            IPAddress ip = IPAddress.Parse(serverIPAddress);
            m_clientSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            try
            {
                m_clientSocket.Connect(new IPEndPoint(ip, port)); //配置服务器IP与端口
                Console.WriteLine("[{0}]> 连接服务器成功: {1}", Name, ReceiveData());
            }
            catch(Exception error)
            {
                Console.WriteLine("[{0}]> 连接服务器失败，{1}", Name, error.Message);
                m_clientSocket = null;
            }
        }

        public bool Login()
        {
            if (m_clientSocket != null)
            {
                try
                {
                    string result = "";
                    JObject json = new JObject();
                    json["name"] = Name;
                    json["action"] = "login";
                    m_clientSocket.Send(Encoding.UTF8.GetBytes(json.ToString()));
                    Console.WriteLine("[{0}]> 登录", Name);
                    result = ReceiveData();
                    dynamic resultJson = JObject.Parse(result);
                    return resultJson["result"].Value == 0;
                }
                catch (Exception err)
                {
                    Console.WriteLine("[{0}]> 登录错误：{1}", Name, err.Message);
                    m_clientSocket.Shutdown(SocketShutdown.Both);
                    m_clientSocket.Close();
                }
            }
            return false;
        }

        public string SendMessage(string toClient, dynamic message)
        {
            string result = "";
            if (m_clientSocket != null)
            {
                try
                {
                    JObject json = new JObject();
                    json["name"] = Name;
                    json["action"] = "send";
                    json["to"] = toClient;
                    json["data"] = message;
                    m_clientSocket.Send(Encoding.UTF8.GetBytes(json.ToString()));
                    Console.WriteLine("[{0}]> 发送消息：{1}", Name, message);
                    result = ReceiveData();
                }
                catch (Exception err)
                {
                    Console.WriteLine("[{0}]> 发送消息错误：{1}", Name, err.Message);
                    m_clientSocket.Shutdown(SocketShutdown.Both);
                    m_clientSocket.Close();
                }
            }
            return result;
        }

        public bool HeartBeat()
        {
            bool alive = false;
            string result = "";
            if (m_clientSocket != null)
            {
                try
                {
                    JObject json = new JObject();
                    json["name"] = Name;
                    json["action"] = "echo";
                    json["data"] = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss");
                    m_clientSocket.Send(Encoding.UTF8.GetBytes(json.ToString()));
                    result = ReceiveData();
                    alive = true;
                }
                catch (Exception err)
                {
                    Console.WriteLine("[{0}]> 发送心跳错误：{1}", Name, err.Message);
                    m_clientSocket.Shutdown(SocketShutdown.Both);
                    m_clientSocket.Close();
                }
            }
            return alive;
        }

        public void Listen()
        {
            if (m_clientSocket != null)
            {
                try
                {
                    while (true)
                    {
                        JObject json = new JObject();
                        json["name"] = Name;
                        json["action"] = "listen";
                        m_clientSocket.Send(Encoding.UTF8.GetBytes(json.ToString()));
                        Console.WriteLine("[{0}]> 发出监听请求（{1}）", Name, json.ToString());
                        var data = ReceiveData();
                        JObject jsonRecv = JObject.Parse(data);
                        Console.WriteLine("[{0}]> 监听消息接收（{1}）", Name, jsonRecv["data"].ToString());
                    }
                }
                catch (Exception err)
                {
                    Console.WriteLine("[{0}]> 监听消息错误：{1}", Name, err.Message);
                    m_clientSocket.Shutdown(SocketShutdown.Both);
                    m_clientSocket.Close();
                }
            }
        }

        public string SingleListen()
        {
            string result = null;
            if (m_clientSocket != null)
            {
                try
                {
                    JObject json = new JObject();
                    json["name"] = Name;
                    json["action"] = "listen";
                    m_clientSocket.Send(Encoding.UTF8.GetBytes(json.ToString()));
                    Console.WriteLine("[{0}]> 发出监听请求（{1}）", Name, json.ToString());
                    result = ReceiveData();
                    Console.WriteLine("[{0}]> 监听消息接收（{1}）", Name, result);
                }
                catch(Exception err)
                {
                    Console.WriteLine("[{0}]> 监听消息错误：{1}", Name, err.Message);
                    m_clientSocket.Shutdown(SocketShutdown.Both);
                    m_clientSocket.Close();
                }
            }
            return result;
        }

        public void CloseConnection()
        {
            if (m_clientSocket != null)
            {
                m_clientSocket.Close();
                m_clientSocket = null;
            }
        }

        internal string ReceiveData()
        {
            string result = "";
            try
            {
                byte[] buffer = new byte[100 * 1024];
                int receiveLength = m_clientSocket.Receive(buffer);
                result = Encoding.UTF8.GetString(buffer, 0, receiveLength);
                Console.WriteLine("[{0}]> 接收服务器消息：{1}", Name, result);
            }
            catch (Exception err)
            {
                Console.WriteLine("[{0}]> 接收消息错误：{1}", Name, err.Message);
            }
            return result;
        }

        public string Name { get; internal set; }
        internal Socket m_clientSocket;
    }
}
